src/NuGet.Core/NuGet.Packaging/PackageCreation/Utility/NuGetFrameworkNameUtility.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Runtime.Versioning;
+
+using NuGet.Frameworks;
+
+namespace NuGet.Packaging
+{
+    public static class NuGetFrameworkNameUtility
+    {
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1021:AvoidOutParameters", MessageId = "1#")]
+        public static NuGetFramework ParseFrameworkNameFromFilePath(string filePath, out string effectivePath)
+        {
+            foreach (string knownFolder in PackagingConstants.Folders.Known)
+            {
+                string folderPrefix = knownFolder + System.IO.Path.DirectorySeparatorChar;
+                if (filePath.Length > folderPrefix.Length &&
+                    filePath.StartsWith(folderPrefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    string frameworkPart = filePath.Substring(folderPrefix.Length);
+
+                    try
+                    {
+                        return NuGetFrameworkNameUtility.ParseFrameworkFolderName(
+                            frameworkPart,
+                            strictParsing: knownFolder == PackagingConstants.Folders.Lib,
+                            effectivePath: out effectivePath);
+                    }
+                    catch (ArgumentException)
+                    {
+                        // if the parsing fails, we treat it as if this file
+                        // doesn't have target framework.
+                        effectivePath = frameworkPart;
+                        return null;
+                    }
+                }
+
+            }
+
+            effectivePath = filePath;
+            return null;
+        }
+
+        /// <summary>
+        /// Parses the specified string into FrameworkName object.
+        /// </summary>
+        /// <param name="path">The string to be parse.</param>
+        /// <param name="strictParsing">if set to <c>true</c>, parse the first folder of path even if it is unrecognized framework.</param>
+        /// <param name="effectivePath">returns the path after the parsed target framework</param>
+        /// <returns></returns>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1021:AvoidOutParameters", MessageId = "2#")]
```

Comment:
Same here, I'm not sure C1021 is enabled in our current analyzers.

Reply:
In fact, we can use a value tuple instead of out params cause they are more explicit.